Post Date: Allow user to pick Site Default, a suggested date format, or a custom date format

noisysocks opened the pull request:
## Description

Closes https://github.com/WordPress/gutenberg/issues/38751.

https://user-images.githubusercontent.com/612155/157785443-17bec04b-f7a6-4113-a66c-22d63697dc64.mp4

Changes the Post Date block so that you can choose between using the site's default date format, a suggested format, or a completely custom format.

## How I arrived at the suggested date formats

_Which formats should we suggest?_ is a difficult question as different locales have different date formatting conventions.

In my opinion there are a few requirements:

1. Users should see suggestions that are appropriate for their locale.
2. We shouldn't overwhelm users with too many options.
3. Theme developers should be able to do basically whatever they want.

We can meet requirement 3 by allowing the theme developer to type any arbitrary date format into a text field.

If we pick a small set of suggested formats, as in https://github.com/WordPress/gutenberg/issues/38751#issuecomment-1039113532, then we will meet requirement 2 but struggle to meet requirement 1. This is because any list we come up with is bound to not be appropriate in some locales. For example, 4/20/22 is typical in US English but 20/04/2022 is typical elsewhere. Or, "April 20, 2022" makes sense in US English but this would be "20 April 2022" in UK English and "20. April 2022" in German.

We could flood the dropdown with hundreds of different options so as to cover all locales and meet requirement 1, but then this would be an overwhelming amount of options and therefore not meet requirement 2.

So what I've opted to do is to start from four categories: short, medium, long, and full. This is inspired by [what Apple has in System Preferences](https://user-images.githubusercontent.com/612155/156981568-28918f2e-65b7-4bcb-bea7-e713b9901b5a.png). For the short, medium, and long categories, I added the option of showing the time as well. This gives seven options. Each option is localised depending on the site's language. That is, the suggested formats will change depending on what language WordPress is set to.

| Option | US English `en_US` | UK English `en_GB` | German `de` |
| - | - | - | - |
| Short | 4/20/2022 | 20/04/2022 | 20.04.2022 |
| Medium | Apr 20, 2022 | 20 Apr 2022 | 20. Apr 2022 |
| Long | April 20, 2022 | 20 April 2022 | 20. April 2022 |
| Full | Wednesday, April 20, 2022 | Wednesday, 20 April 2022 | Mittwoch, 20. April 2022 |
| Short + time | 4/20/2022 4:20 AM | 20/04/2022 4:20 am | 20.04.2022 04:20 |
| Medium + time | Apr 20, 2022 4:20 AM | 20 Apr 2022 4:20 am | 20. Apr 2022 04:20 |
| Long + time | April 20, 2022 4:20 AM | 20 April 2022 4:20 am | 20. April 2022 04:20 |
| ISO | 2022-04-20 | 2022-04-20 | 2022-04-20 |

## Notes

- A `null` `attributes.format` is what denotes that the user has selected _Site default_. This is the default.
- Existing Post Date blocks will remain as they are, even if they use the site's default format. I don't think we can properly migrate this as we don't know whether `attributes.format === siteFormat` because the user intended that or if it's just a coincidence.
- We don't need to access the `@wordpress/date` settings as we can get everything we need from `/wp/v2/settings` (the `site` entity).
- ~I need to update the Comment Date block with the same changes. (Maybe the dropdown should be a block editor component? (Or a block supports? 😬))~
- I removed a bunch of unnecessary string/array manipulation from the `is12Hour` calculation and fixed the logic of this. Previously it was only looking for `a` but technically a 12 hour format string could just be e.g. `g:i` with no `a`.

## Testing Instructions

1. Open the site editor.
6. Find a Post Date block and go into its block settings.
7. Experiment with different _Format settings_.
8. If _Site default_ is selected, check that changing the date setting in _Options → General_ changes how the Post Date appears in the frontend.
9. If _Custom_ is selected, check that you can enter a custom date format string.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -→
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -→
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -→
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -→
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -→
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -→
- [x] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

noisysocks committed:
Post Date: Allow user to pick Site Default, or a suggested date format

andrewserong commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -35,30 +35,39 @@ export default function PostDateEdit( {
-	const resolvedFormat = format || siteFormat || settings.formats.date;
+	// To know if the time format is a 12 hour time, look for any of the 12 hour
+	// format characters: 'a', 'A', 'g', and 'h'. The character must be
+	// unescaped, i.e. not preceded by a '\'. Coincidentally, 'aAgh' is how I
+	// feel when working with regular expressions.
```
I hope this comment makes it when this PR eventually lands 😆

ramonjd replied:
Please leave this comment in.

noisysocks replied:
I couldn't possibly remove it.

talldan commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -35,30 +35,39 @@ export default function PostDateEdit( {
+				dateI18n( siteFormat, date )
+			),
+		},
+		...[
+			_x( 'M jS', 'date format option' ),
+			_x( 'M j, Y', 'date format option' ),
+			_x( 'M j, Y, h:i A', 'date format option' ),
+			_x( 'F j, Y', 'date format option' ),
+			_x( 'd/m/Y', 'date format option' ),
+		].map( ( suggestedFormat ) => ( {
+			key: suggestedFormat,
+			name: dateI18n( suggestedFormat, date ),
+		} ) ),
```
Probably an issue in trunk too, so it might be something to fix separately, but thought I'd mention it as it might be a fix we can roll in to this PR.

The labels could be improved for screenreaders. Both 'Feb 25, 2022' and 'Februrary 25, 2022' are announced as 'February 25th 2022' using Voiceover. I it does some normalisation of dates. I'm not sure if we can add some screenreader text into the `CustomSelectControl` items that also contains the format.

Let me know if you want me to create a separate issue.

noisysocks replied:
Oh interesting. That's tricky. I'm not entirely sure what the expectations would be for folks using screen readers. Go ahead and create a new issue so that we can ask for feedback on that from a11y folks.

talldan replied:
Here's a new issue - https://github.com/WordPress/gutenberg/issues/39278

Testing again, and there's also no way to tell which item in the dropdown is currently checked. 😞

andrewserong replied:
I just had a quick look at the SelectControl component, and it looks like it doesn't currently set the `selected` attribute on option elements ([here](https://github.com/wordpress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/packages/components/src/select-control/index.tsx#L152)). I was wondering if that could be the culprit? (Probably something to look at separately to this PR?)

noisysocks committed:
Update suggested formats to be more localizable

noisysocks committed:
Add support for custom formats

noisysocks committed:
Move is12Hour check to function

noisysocks commented:
I thought some more about what formats to suggest to the user (see _How I arrived at the suggested date formats_ in the description) and made some changes there. I also added the ability to set a custom format (see updated video in the description) which I think is necessary as we can't cover every date format that a theme author might want in the dropdown.

Keen for feedback! Particularly on my thinking around what date formats to show. Internationalisation makes it very tricky.

andrewserong commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -155,7 +138,114 @@ export default function PostDateEdit( {
+	return /(?:^|[^\\])[aAgh]/.test( format );
+}
+
+function FormatSettings( { date, format, siteFormat, setFormat } ) {
```
> I need to update the Comment Date block with the same changes. (Maybe the dropdown should be a block editor component? (Or a block supports? 😬))

Would it be worth moving this to the `block-editor` package so that the component can be re-used with the Comment Date block? I think that would make good sense for re-usability before exploring a date block support (which isn't a bad idea, but we might want to see if there's a good case for additional blocks before going too far further down that path?)

noisysocks replied:
Split it out into `DateFormatControl` in 102dd3867ef3e8327fd9e91616116e32b03f27ee 👍

andrewserong commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -155,7 +138,114 @@ export default function PostDateEdit( {
 	);
 }
+
+function is12HourFormat( format ) {
```
Is it worth adding a unit test for this function, since it's a small refactor?

noisysocks replied:
Sure is 👍 5f2851c8826c54945861a53b1f459af53b95a8f0

andrewserong reviewed (commented):
This is looking great to me @noisysocks, thanks for the excellent PR description and explanation of the decisions involved! I particularly like being able to edit the custom date format in realtime in the editor UI. While this is an advanced control, I agree, it'd be really great to be able to open up this level of control to themers, and keeping it hidden behind the Custom setting means that we're not unintentionally exposing it to novice users 👍

![image](https://user-images.githubusercontent.com/14988353/156979206-f2693b84-80ce-4be7-9bb8-1f4d623a5ac8.png)

In a future follow-up, it'd be fun for us to explore a WYSIWYG interface for dragging/dropping formatting for dates, but that exploration is for another time 😃

One thing I noticed, since you mentioned i18n, is that as an Australian user of my US English test site, it isn't immediately noticeable from the formatting suggestions if a number is day or month. I'm not sure if there's anything we can really do easily for it at this stage, but just thought I'd mention it:

![image](https://user-images.githubusercontent.com/14988353/156979166-dabae7ed-f6d6-4450-9623-acf1bae8cec8.png)

noisysocks committed:
Distinguish between 'long' and 'full' like Apple does

noisysocks commented:
> In a future follow-up, it'd be fun for us to explore a WYSIWYG interface for dragging/dropping formatting for dates, but that exploration is for another time

Yeah, this is probably the only way to not have to make any compromises. It's obviously a bit of work to implement such an interface, though. I'm not sure if the cost vs benefit stacks up just yet.

> One thing I noticed, since you mentioned i18n, is that as an Australian user of my US English test site, it isn't immediately noticeable from the formatting suggestions if a number is day or month. I'm not sure if there's anything we can really do easily for it at this stage, but just thought I'd mention it:

We could change the date in the dropdown to use a sample date with a day greater than 12. For example, 20221-01-25. I'm not sure if this is more unclear, though. It might lead users to get confused and think that their post's publish date is the sample date. What do you think?

noisysocks commented:
> We could change the date in the dropdown to use a sample date with a day greater than 12. For example, 20221-01-25. I'm not sure if this is more unclear, though. It might lead users to get confused and think that their post's publish date is the sample date. What do you think?

Another reason to do this is that if we're using the post date it's not always clear what the difference is between the medium date which abbreviates the month name and the long date which doesn't. For example, if the post date is in May, it won't be clear because May is already three letters.

I'll switch to using a sample date. We can change our minds later.

noisysocks committed:
Use a sample date in the dropdown so as to better illustrate the formats

andrewserong commented:
>  It's obviously a bit of work to implement such an interface, though. I'm not sure if the cost vs benefit stacks up just yet.

Agreed, for the moment we probably need a stronger use case for date formats being frequently edited to make it a worthwhile investment of time in building it.

> I'll switch to using a sample date. We can change our minds later.

I think the sample date approach sounds like a good way forward to me!

noisysocks committed:
Split out DateFormatControl and update the Comment Date block

noisysocks committed:
Add basic tests for is12HourFormat

noisysocks commented:
I moved the dropdown into its own component and am using that in the Comment Date block now too. Easiest way to test that is to edit a post, edit the template (Post inspector → Template → Edit), and then insert a Comments Query Loop block.

andrewserong commented on packages/block-editor/src/components/date-format-control/README.md:
```diff
@@ -0,0 +1,64 @@
+
+All date format strings should be in the format accepted by by the (`dateI18n`
+function in
+`@wordpress/date`)[https://github.com/WordPress/gutenberg/tree/trunk/packages/date#datei18n].
```
Tiny formatting issue: looks like this markdown syntax is the wrong way around 😀 (should be `[]` followed by `()`)

![image](https://user-images.githubusercontent.com/14988353/157353756-0dcbacbd-8451-44b9-853f-9824c3a1809c.png)

noisysocks replied:
Gah, every time 😂

andrewserong commented on packages/block-editor/src/components/date-format-control/index.js:
```diff
@@ -0,0 +1,140 @@
+	// 2022) in German (de). The resultant array is de-duplicated as some
+	// languages will use the same format string for short, medium, and long
+	// formats.
+	const suggestedFormats = uniq( [
```
Not a blocker for now, but in the future we might want to allow an extra param on this component to either replace or append to this list of suggested formats. I reckon we could leave that to a follow-up if and when we need it, though!

Also, just double-checking, is the `uniq` call here just in case the translated / localized strings winds up with duplicates?

noisysocks replied:
Yeah, I'm not sure how best to make this component "generic" so I figure keep the API minimal for now and it can evolve with time.

Correct. I anticipate that some languages don't have e.g. a distinction between medium dates (Jan 5) and long dates (January 5) and so best to de-dupe the list of localised strings. At the very least, I know most Asian languages use YYYY-MM-DD which is a format we already have by default.

ramonjd replied:
> Also, just double-checking, is the uniq call here just in case the translated / localized strings winds up with duplicates?

I was just coming to this line to ask the same question. 😄  Is it worth documenting this inline somewhere?

andrewserong reviewed (approved):
This is testing really well for me @noisysocks, thanks for re-arranging things and adding in the extra test! Just found a tiny formatting issue in the documentation where I've added a comment.

✅ Works with site's datetime formatting settings
✅ Post Date and Comment Date blocks work correctly with the new component
✅ Suggested and custom date formats work as expected, and escaped characters can be used to inject a word like "at":

![image](https://user-images.githubusercontent.com/14988353/157354743-ddcf814f-a185-4987-b454-5d8f422b2fed.png)

✅ Custom date formats are matched against the list of suggested formats, so on editor reload if there's a match, the suggested format is highlighted if there's a match, and if not, the custom field is pre-selected

Thanks again for the detailed testing notes / explanations, the code change LGTM! It'll be good to get some design feedback on the suggested formats, too — they look like a good starting point to me, though I was wondering what the preference would be between capitalised / not capitalised AM/PM / am/pm?

noisysocks committed:
Fix markdown link formatting

noisysocks commented:
Thanks for the review! I definitely want to get some design feedback on my thinking in the description and the list of suggested formats before merging.

> though I was wondering what the preference would be between capitalised / not capitalised AM/PM / am/pm?

From my research this seems to be locale specific. Or at least that's how Apple has implemented it. American English uses "AM", UK English uses "am", and most other languages use 24 hour time.

I appreciate that that's all pretty academic though. It's difficult to strike a balance between having suggestions that are _correct for the site's locale_ and _useful for theme developers_ while also not having hundreds of options in the dropdown. Again, feedback welcome 😅

andrewserong commented:
> Again, feedback welcome 

Thanks for the explanations, I didn't realise it varied so much by locale! That all sounds reasonable to me, and I quite like the defaults you've added. Let's see what the design experts think 😀

kjellr commented:
I think all the decisions here make sense. I like the presets, the idea to leverage localization, as well as the use of a sample date to ensure that the previews are clear. 

I like the Custom functionality too, but I'm wondering if we have that UI pattern (selecting custom in a dropdown to show an additional field) in place for other blocks anymore? I think the closest we have at the moment is the custom field for font size, which uses a slightly different paradigm: 

![type](https://user-images.githubusercontent.com/1202812/157446512-1995e471-d22b-46a0-83d5-a1a2f30d4cf0.gif)

That doesn't feel quite right here though, since I think it would over-emphasize the custom option. I'd like one more designer to take a look and weigh in. @jasmussen or @jameskoster perhaps?

jasmussen commented:
Took it for a spin, here are the test instructions in GIF form:
![date](https://user-images.githubusercontent.com/1204802/157457296-4d049a87-6e05-4cb0-95b1-a7c72a16cd90.gif)

Very nice, works well. A few thoughts; to echo Kjell, the "Custom" one feels a bit weird since it's the only dropdown option that shows a separate input field when selected. Reusing the same format as is used for font size would be a nice bit of consistency. 

The bigger problem with "Custom", I feel, is its juxtaposition against the `Site default` option at the top; to an extent, these date formats are ALL custom, because they are detached from the site-wide setting. I wonder if we should do something similar as to what's done with the Layout option, and have an a toggle to inherit the default date format:
![inherit](https://user-images.githubusercontent.com/1204802/157457987-66aa6409-be0d-4958-a8e7-4e85a40f458b.gif)

If we do that, we should _show_ that default layout format in help text below the toggle, when it's enabled. I also wonder if we can find a different term than "Custom", possibly just be more verbose? "Custom date format"?

We have a tendency to proliferate collapsible panels and use them like subheadings. For such a single purpose block, I don't think we need to. Instead of a panel for "Format settings" with a single option, and a panel for "Link settings" with a single toggle, could we just have a single "Settings" panel, that included both date format and link settings? "Link to post" feels plenty descriptive on its own.

Finally, this is a long list of options:
<img width="294" alt="Screenshot 2022-03-09 at 15 13 55" src="https://user-images.githubusercontent.com/1204802/157458888-f7545763-ad23-4b7a-9999-00b47a39b0d1.png">

Blink, and you'll miss that the 3 of them include the clock timestamp. Should we reduce to, say, 5 options? Or if we do need this many options, we might want to look at a custom dropdown so we could add separators and/or subheadings.

jameskoster commented:
> these are ALL and detached from the default.

I was going to say exactly this. If it's not bad practise to only have two options in a segmented control, perhaps that would be an option versus the toggle?

<img width="1222" alt="Screenshot 2022-03-09 at 14 19 27" src="https://user-images.githubusercontent.com/846565/157460025-e57fc710-b101-4981-93c3-0be2b89cb376.png">

noisysocks commented:
Great suggestions ❤️ thank you!

ramonjd commented on packages/block-library/src/post-date/edit.js:
```diff
@@ -33,47 +33,30 @@ export default function PostDateEdit( {
 	let postDate = date ? (
 		<time dateTime={ dateI18n( 'c', date ) } ref={ timeRef }>
-			{ dateI18n( resolvedFormat, date ) }
+			{ dateI18n( format || siteFormat, date ) }
```
I think we might need a hook somewhere to ensure `siteFormat` is available.

For me at least, there seems to be a race condition whereby, if I refresh the Post Editor with a Post Date block and the site default selected, both `format` and `siteFormat` are `undefined`. 

This is only in the Post Editor. The Site Editor works as expected.

The result is, when refreshing the page, [format() complains that it cannot find a length](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/packages/date/src/index.js#L387).

Adding a new Post Date block reveals that the value is indeed returned at some point after refresh.

noisysocks replied:
Ah bugger. I guess it happens while the site entity is loading.

I added a fallback to the `@wordpress/date` format in b3fcfdaa3a114b1c973a15d45952cea71c0bfa3c.

Let me know if that fixes it for you.

ramonjd commented on packages/block-editor/src/components/date-format-control/index.js:
```diff
@@ -0,0 +1,140 @@
+		_x( 'n/j/Y g:i A', 'short date format with time' ),
+		_x( 'M j, Y g:i A', 'medium date format with time' ),
+		_x( 'F j, Y g:i A', 'long date format with time' ),
+		'Y-m-d',
```
A question about the order:

I see the more elaborate datetime formats are grouped.

In no way a biggie, but would it make sense to bump the shorthand format at the bottom `YYYY-MM-DD` up a bit so it sits cosy beneath its slashed cousin? 

That is, from


<img width="278" alt="Screen Shot 2022-03-10 at 11 07 37 am" src="https://user-images.githubusercontent.com/6458278/157563179-3018f132-3227-4faa-98fd-ff2113724c36.png">

to 

<img width="277" alt="Screen Shot 2022-03-10 at 11 14 36 am" src="https://user-images.githubusercontent.com/6458278/157563171-945410b3-aacc-43bf-83d5-70248b63edf8.png">

noisysocks replied:
I played around with the order a little bit in ba74a6b3d0f54f15af939ac0aa9e87cdbf0930f9. I think it makes more sense now.

ramonjd commented:
Fantastic work! Thank you for taking this so far and in such a positive direction. I ❤️ the custom field.

✅  Post date is displayed in the site and post editor as expected. 
✅  On the frontend, the post date reflect my selected format. Where I've modified it in the Post Editor, it saves the format for the particular post only (and not everywhere else).
✅  When I change the site default in site settings, it's reflected in the drop down list.
❓ Just had [one observation](https://github.com/WordPress/gutenberg/pull/39209#discussion_r823227200) about `siteFormat` not being immediately available in the Post Editor.

noisysocks committed:
Combine 'Format settings' and 'Link settings' into single 'Settings' panel

noisysocks committed:
WIP

noisysocks committed:
Use Default / Custom radio control

Re-jig DateFormatControl into DateFormatPicker. Users now select either
Default or Custom in a ratio control and, then, if they select Custom,
they can select a suggested date format or enter a manual one.

noisysocks committed:
Simplify suggested date formats

noisysocks committed:
Use date and time formats from @wordpress/date as a fallback

noisysocks commented:
OK, I've redone this to be close to the design in https://github.com/WordPress/gutenberg/pull/39209#issuecomment-1062969878.

It's not an exact match as there's some differences between the mockup @jameskoster shared and how `CustomSelectControl` looks. Probably we should update that component at some point in the future as I think @jameskoster's mockup looks nicer—but not right now! 😀

Check out the PR description for an updated screen recording of how it looks.

talldan commented on packages/block-editor/src/components/date-format-picker/index.js:
```diff
@@ -0,0 +1,170 @@
+		name: __( 'Other' ),
+		className:
+			'block-editor-date-format-picker__custom-format-select-control__other-option',
+		__experimentalHint: __( 'Choose your own date string' ),
```
This is a nitpick, but also a reaction from using it for the first time. The word 'choose' made me think I would be presented with another dropdown. 'String' is also very computer science-y.

A suggestion would be 'Enter your own date format'.

noisysocks replied:
Ooh I like that 👍 f6e57fe264

talldan commented on packages/block-editor/src/components/date-format-picker/index.js:
```diff
@@ -0,0 +1,170 @@
+				value={ format ? 'custom' : 'default' }
+				onChange={ ( value ) => {
+					if ( value === 'default' ) {
+						onChange( null );
```
It's interesting that it gets set to `null`. That possibly presents the implementor of the component with a gotcha, they have to ensure they implement a fallback to the `defaultFormat` themselves.

I feel like it might be easier to implement if `format` acted more like the `value` of a controlled input and `onChange` always provided the selected format.

An option could be to not show the default format in the `CustomSelectControl` as an option, then the logic within the component becomes a bit simpler, and I think it also removes some overhead for both the implementor and the end user (because I don't they need two different ways to select it).

It is a design decision though.

noisysocks replied:
Maybe I am not understanding you properly (it's getting late on a Friday lol) but I think this is right because what you're indicating when you choose _Default_ is that _I don't have a preference_. It's similar to the layout controls where one of the options is to choose _Inherit default layout_. We're using `null` to indicate _nothing_ which is semantically correct.

noisysocks replied:
To put it another way I think there's a meaningful difference between selecting _Default_ and getting `M i, Y` and selecting _Custom_ and choosing `M i, Y`. The latter wouldn't change if you update the setting in Options → General.

talldan replied:
Ah, yes you're right, that wouldn't work if the component always returns the default as its value. 👍 

It's a shame it requires using `null` to signal that the default should be used, but I can't think of a better way. You've probably thought about this a lot too.

Maybe a little bit of extra detail could be added to the doc block for `props.format` to document that the `null` value means the default has been chosen.

noisysocks replied:
👍 good idea ce9835d152

talldan commented on packages/block-editor/src/components/index.js:
```diff
@@ -42,6 +42,7 @@ export {
 export { default as ColorPalette } from './color-palette';
 export { default as ColorPaletteControl } from './color-palette/control';
 export { default as ContrastChecker } from './contrast-checker';
+export { default as DateFormatPicker } from './date-format-picker';
```
This seems like a very generic component, so I wonder if `block-editor` is the right package.

Most of the other block editor components are either connected to data or very block specific.

noisysocks replied:
Yeah I umm'd and ahh'd about `@wordpress/block-editor` versus `@wordpress/components`. It only depends on `@wordpress/date` so we can put it anywhere, really. What do you think?

talldan replied:
I'm leaning towards `components`. I think it's ok to ship it in block-editor in this PR, but maybe don't make it stable to start with, so that it's still ok to move it later.

As a components package maintainer, it'd be good to get some thoughts from @ciampo about this one.

ciampo replied:
Thank you for the ping!

TL;DR: We could consider moving it to the `@wordpress/components` package, although I suggest doing that in a separate PR. I 100% with @talldan 's suggestion about marking the component _unstable_ for the time being.

---

A few more detailed considerations:

- it's true that this component is generic for the `block-editor` package, so it would make sense to move it somewhere else. In case we moved it to `@wordpress/components`, though, we'd need to tweak the APIs a bit to make even more generic and decouple the component a bit more from the block editor (e.g. `siteFormat` could probably become `defaultFormat` ?)
- in case we moved the component to `@wordpress/components`, we'd need to refactor it:
    - from JS to TypeScript
    - from sass to CSS-in-JS (currently emotion)
    - add unit tests
    - add Storybook examples
    - ...and more in general, follow these [guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md)

So probably the best thing is to introduce the component in the `block-editor` package as experimental, then move it to `@wordpress/components` and refactor it as describe above in a separate, follow-up PR.

Also cc'ing @mirka (who also helps maintaining the components package) for any additional thoughts.

noisysocks replied:
Thanks @ciampo!

> So probably the best thing is to introduce the component in the block-editor package as experimental, then move it to @wordpress/components and refactor it as describe above in a separate, follow-up PR.

I shall do exactly this.

talldan commented on packages/block-editor/src/components/date-format-picker/index.js:
```diff
@@ -0,0 +1,170 @@
+				/>
+			</BaseControl>
+			{ isOther && (
+				<TextControl
+					label={ __( 'Format string' ) }
+					hideLabelFromVision
+					help={ createInterpolateElement(
+						__(
+							'Enter a date or time <Link>format string</Link>.'
+						),
+						{
+							Link: (
+								<ExternalLink
+									href={ __(
+										'https://wordpress.org/support/article/formatting-date-and-time/'
+									) }
+								/>
+							),
+						}
+					) }
+					value={ format }
+					onChange={ ( value ) => onChange( value ) }
+				/>
```
Would be good to revise the use of the word 'string' here as well, IMO.

noisysocks replied:
Hm I want to use the same nomenclature that we use in [the docs that these two words link to](https://wordpress.org/support/article/formatting-date-and-time/), though.

> Notice the string of characters next to each Date and Time formatting in screenshot. This string is called a format string. Each letter represents specific part of Date or Time.

talldan replied:
This documentation also straight away mentions PHP functions and the PHP programming language. I don't think most end users should be exposed to this kind of stuff, so my opinion is that the docs are wrong here too. They seem like technical docs.

noisysocks replied:
True, the docs could use a bit of work. It's what we link to from Options → General. I want to be consistent there.

To be honest this is a very technical feature and I don't think there's any point shying away from that. Maybe one day we will build a drag-and-drop style interface like @andrewserong suggests and hide the technical nature of it, but until then I think we just have to wear it. (And hide it in a dropdown where only power users will look, which we've done 😀)

talldan reviewed (commented):
This new design works really well, IMO. 

The only thing that I'm not sure about is that it requires selecting 'Custom' and then 'Other' for a custom date format, and I've always been told not to use the word 'Other' for these kinds of things. 

When testing using a screen reader it felt like the word 'custom' would be best used to describe the text input field (as it is in the WordPress settings), but then I don't know what the alternative to 'Default' would be if 'Custom' weren't the word used there.

Since I don't have any good suggestions, maybe best to leave it as is. 

Looks like a really solid implementation too. Just a few comments.

noisysocks commented:
@talldan: I noticed that I intuitively reached for the word "manual" when writing the git commit. What do you think of "manual" instead of "other"?

noisysocks committed:
Use 'Enter your own date format'

talldan commented:
I was thinking there might be a better word to describe the second option in the toggle group, and then 'custom' could be used for the last option in the dropdown and the text input. I was trying to think of a word that means "Here are some convenient date format options we've picked for you". Some words that spring to mind are 'Common', 'Standard', 'Curated'. Now I'm not sure though. I've thought about it too much and I've lost the ability to judge. 😄 

I think that part is good to leave as it is, my comments were more thinking aloud rather than a voicing a concern.

Of all the feedback I left, I feel most strongly about removing the word 'string' as I don't think all users would know what this means.

noisysocks commented:
> I think that part is good to leave as it is, my comments were more thinking aloud rather than a voicing a concern.

👍😀 language is hard. We can easily change the copy later if we think of a great idea.

> Of all the feedback I left, I feel most strongly about removing the word 'string' as I don't think all users would know what this means.

I changed one of the occurrences but left the other as I think it's nice that that occurrence of "format string" is aligned with what it's called in [the documentation](https://wordpress.org/support/article/formatting-date-and-time/) that it links to. As in: you see "format string", it's blue, you think "what's that?", you click it, now you learn what a "format string" is. The Venn diagram of "people who are confused by 'format string'" and "people who I think should click on the link" is a circle.

noisysocks committed:
Explain what null means in the doc comment

jameskoster commented:
Yeah the labelling is hard. It may actually make more sense to swap the 'Other' and 'Custom' labels around. The result being that you can let the format inherit the default setting, or choose _another_.

Then if you select Other you can choose from presets, or create something custom.

Of course the problem then is that 'Other' is a bit of a strange label as pointed out by @talldan. On that point, Joen's original instinct to use a toggle might be the best one because it circumvents the need to surface 'Other' as a label, while communicating that anything besides the default is a departure from the global setting.

<img width="916" alt="Screenshot 2022-03-11 at 10 28 30" src="https://user-images.githubusercontent.com/846565/157850245-fa93a660-edeb-4fd9-b556-28a00f729217.png">

The only thing we're missing then is a preview of the custom string in the control. But if that will render on the canvas it's probably not a big deal.

What do y'all think?

jameskoster commented:
Separately, now that we've combined all the settings I think I spotted a bug. While editing something like the single post template in the site editor the link toggle label seems to be broken:

<img width="251" alt="Screenshot 2022-03-11 at 10 35 57" src="https://user-images.githubusercontent.com/846565/157851131-e3807575-b69d-4109-957e-bd17cab1016d.png">

I presume this is because it's looking for a post type or title that is not provided.

jasmussen commented:
I like it, Jay! Instead of `Default — March 27, 2022`, I wonder if `Default (March 27, 2022)` would read better?

noisysocks committed:
Use ToggleControl for selecting a default date

noisysocks committed:
Use __experimental like a coward

noisysocks commented:
Alrighty I swapped the toggle group for a regular toggle component, like in https://github.com/WordPress/gutenberg/pull/39209#issuecomment-1064985823. I think it definitely is clearer now that "custom" means only one thing. I'll update the video in the description. 

@jameskoster: In the design you had two labels for the toggle but unfortunately that's not how our toggle component works. Toggles only have one label which is on the right. I could add text above the toggle but it wouldn't work like a regular label (clicking it should focus the field) and it'd probably confuse assistive technology users.

I worked around this by using "Default format (March 27, 2022)" for the copy.

<img width="296" alt="Screen Shot 2022-03-17 at 15 42 54" src="https://user-images.githubusercontent.com/612155/158739001-d3091b97-2655-4e08-b0af-219f1fb2e0c9.png">

We could also put the date on a new line which I think might look a little nicer.

<img width="299" alt="Screen Shot 2022-03-17 at 15 45 38" src="https://user-images.githubusercontent.com/612155/158739010-199e9a55-a5b3-4ff9-94da-20f487d5980b.png">

Happy to hear other suggestions.

jasmussen commented:
> Happy to hear other suggestions.

The toggle should be able to have helptext on an entirely new line, like Drop Cap right?
<img width="239" alt="Screenshot 2022-03-17 at 09 58 31" src="https://user-images.githubusercontent.com/1204802/158773887-0744c30b-bf15-4674-9492-6b4ae4f05589.png">

☝️ note there's something off about the spacing there, but that's a separate issue.

noisysocks committed:
Move example of default format to new line

noisysocks commented:
Yes correct!

I think it feels a little disconnected though:

<img width="296" alt="Screen Shot 2022-03-18 at 11 20 02" src="https://user-images.githubusercontent.com/612155/158914318-da17113e-48d8-4aa5-82e1-c232debeb571.png">

I don't think there's a problem with the spacing, it's just that the help text usually shows a sentence or two of copy:

<img width="298" alt="Screen Shot 2022-03-18 at 11 21 15" src="https://user-images.githubusercontent.com/612155/158914438-cc9e75f2-6fe7-49da-ad61-fd3e0012439a.png">

I pushed 08d3e41 which moves the date to a new line under the label, which is my preference:

<img width="293" alt="Screen Shot 2022-03-18 at 11 41 49" src="https://user-images.githubusercontent.com/612155/158915956-6e2367ec-f988-4289-a1a0-4cf7c1589ec3.png">

I'd love to get this merged and do any further iteration in follow-up PRs! What we have here is a lot better than what's in `trunk` 😀

noisysocks committed:
Fix 'Link to post' label

noisysocks commented:
> Separately, now that we've combined all the settings I think I spotted a bug. While editing something like the single post template in the site editor the link toggle label seems to be broken:
> 
> <img alt="Screenshot 2022-03-11 at 10 35 57" width="251" src="https://user-images.githubusercontent.com/846565/157851131-e3807575-b69d-4109-957e-bd17cab1016d.png">
> 
> I presume this is because it's looking for a post type or title that is not provided.

Good catch. Fixed in d6cc73e135.

talldan reviewed (approved):
It works really well, great job iterating here.

This is a good improvement over what's in `trunk`, so I really think it's worth merging and then following up with any further design changes.

As an aside, I noticed another bug with the formats while testing, but it seems like an existing issue with the `dateI18n` utility rather than anything introduced here, so I've made an issue for it - #39567.

jasmussen commented:
Nice one, thanks for working on this.

Let's remember to follow up on the help text, so this isn't the only place where it's used like this:
<img width="304" alt="Screenshot 2022-03-18 at 09 30 49" src="https://user-images.githubusercontent.com/1204802/158965354-7fd7eb85-d2dc-42b0-a383-42dbbeffbdb5.png">

For lack of larger improvements such as those Jay suggested, we could start with something simple, like this:

<img width="299" alt="Screenshot 2022-03-18 at 09 35 24" src="https://user-images.githubusercontent.com/1204802/158965403-dfe3a08b-7ba7-45e3-a21d-e85928eea229.png">

CC: @ciampo, I think there's some margin between toggle and help text we need to tweak so it's smaller. More specifically, when a toggle has helptext, the toggle itself doesn't need to have bottom margin, it appears, since the helptext itself has that.

ciampo commented:
> CC: @ciampo, I think there's some margin between toggle and help text we need to tweak so it's smaller. More specifically, when a toggle has helptext, the toggle itself doesn't need to have bottom margin, it appears, since the helptext itself has that.

Good catch! @mirka , can this also be part of the recent style "normalisation" treatment that you're applying across components?

mirka commented:
> can this also be part of the recent style "normalisation" treatment that you're applying across components?

Yes, that will fit in perfectly with the work in #38730 ✨  @jasmussen I'll ping you on the PR for a check when we do `ToggleControl`.
